### PR TITLE
Fix RStan 2.26 building & Standalone functions

### PR DESCRIPTION
**Pull request by andrjohns**
This PR updates `rstan_config()` to define `USE_STANC3` prior to including stan headers, and also adds functionality for building packages with stan models of just standalone functions

### CODE REVIEW

File: inst/include/sys/Makevars [hunk 1]
```diff
@@ -1,6 +1,6 @@
 STANHEADERS_SRC = $(shell "$(R_HOME)/bin$(R_ARCH_BIN)/Rscript" -e "message()" -e "cat(system.file('include', 'src', package = 'StanHeaders', mustWork = TRUE))" -e "message()" | grep "StanHeaders")
 
-PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DBOOST_MATH_OVERFLOW_ERROR_POLICY=errno_on_error
+PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DBOOST_MATH_OVERFLOW_ERROR_POLICY=errno_on_error -DUSE_STANC3
```
Comment (hsbadr) [inline inst/include/sys/Makevars:3]: `USE_STANC3` should only be defined if the C++ code is generated by `stanc3`. This is ok when `rstan 2.26` is released, but the transition to CRAN may require releasing https://github.com/stan-dev/rstan/pull/912 first that's compatible with the older version of `rstan`. So, if `USE_STANC3` is defined, Math v4+ headers will be exposed and break the C++ code generated by `stanc2`.

**Solution:** While the development version of `rstan` defines `USE_STANC3` in the generated C++ code, it's safe to define it here only `if (utils::packageVersion("rstan") >= 2.26)`.

Comment (hsbadr) [reply inst/include/sys/Makevars:3]: @andrjohns A variant of the following should work:
```bash
"$(R_HOME)/bin$(R_ARCH_BIN)/Rscript" -e "cat(ifelse(utils::packageVersion('rstan') >= 2.26, '-DUSE_STANC3', ''))"
```

Comment (hsbadr) [inline inst/include/sys/Makevars:3]:
```suggestion
STANC_FLAGS = $(shell "$(R_HOME)/bin$(R_ARCH_BIN)/Rscript" -e "cat(ifelse(utils::packageVersion('rstan') >= 2.26, '-DUSE_STANC3',''))")
PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DBOOST_MATH_OVERFLOW_ERROR_POLICY=errno_on_error $(STANC_FLAGS)
```

File: inst/include/sys/Makevars.win [hunk 2]
```diff
@@ -1,6 +1,6 @@
 STANHEADERS_SRC = $(shell "$(R_HOME)/bin$(R_ARCH_BIN)/Rscript" -e "message()" -e "cat(system.file('include', 'src', package = 'StanHeaders', mustWork = TRUE))" -e "message()" | grep "StanHeaders")
 
-PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DRCPP_PARALLEL_USE_TBB=1
+PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DRCPP_PARALLEL_USE_TBB=1 -DUSE_STANC3
```
Comment (hsbadr) [inline inst/include/sys/Makevars.win:3]:
```suggestion
STANC_FLAGS = $(shell "$(R_HOME)/bin$(R_ARCH_BIN)/Rscript" -e "cat(ifelse(utils::packageVersion('rstan') >= 2.26, '-DUSE_STANC3',''))")
PKG_CPPFLAGS = -I"../inst/include" -I"$(STANHEADERS_SRC)" -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -DRCPP_PARALLEL_USE_TBB=1 $(STANC_FLAGS)
```

**Comment by andrjohns** [comment]
@hsbadr sorry I forgot about this! I also modified the function called by `rstan_config` to append the rstantools Makevars to the package-included makevars. This way the user just needs to include the rstantools `configure` files, and then we can update the rstantools `Makevars` as needed.

What do you think?

**Comment by hsbadr** [comment]
Thanks @andrjohns! This makes sense to me, particularly because we plan to require using `rstantools` moving forward. I haven't tested this PR with the different versions of `rstan` (CRAN ~ should be ok, transitional 2.26, experimental/development 2.27+). I think it can stay pending until we decide on the best way to address the issues with CRAN submission. In one of the potential solutions, we may be able to redirect the Stan/Math headers from `StanHeaders` to `rstan` in order to fix the circular dependency (`rstan` will be updating the headers directly) but that's still under discussion.

**Comment by andrjohns** [comment]
I've done some basic testing with CRAN rstan + StanHeaders 2.26 and rstan 2.26 + StanHeaders 2.26 and all seems to be working, but will test a bit more over the next couple of days.

In terms of getting things out to CRAN quickly, I'd be leaning towards an `rstantools` implementation like this one, since it requires pretty minimal changes from package maintainers in the first instance and would be ready quickly. But I'm easy either way

**Comment by andrjohns** [comment]
@hsbadr can we merge a fix with just the addition of the `STANC_FLAGS` section (i.e., removing the changes to `rstan_config`)? That would fix a good deal of reverse dependencies for 2.26 at the moment

**Comment by hsbadr** [comment]
> @hsbadr can we merge a fix with just the addition of the `STANC_FLAGS` section (i.e., removing the changes to `rstan_config`)? That would fix a good deal of reverse dependencies for 2.26 at the moment

Sounds good to me. The major challenge with CRAN releases is the backward incompatibility at C++ level. Some packages either link directly to `StanHeaders` using their pre-generated C++ code or don't rerun/update Stan to C++ code via `rstantools`. This requires patching the headers, which in turn breaks many packages especially because we can't release both `StanHeaders` and `rstan` at the same time. To address this (completely for all packages that use `rstantools` & partially for the minority of the other packages), we're testing a redirection of the headers to `rstan` (and reversing the dependency) in a few minor releases. This may require a legal advice (licensing) though. We'll likely open this for discussion on the forums.

**Comment by andrjohns** [comment]
Great! I've made those changes. Would we be able to get this out to CRAN?

**Review by hsbadr — APPROVED** [review]

**Comment by hsbadr** [comment]
> Great! I've made those changes. Would we be able to get this out to CRAN?

I've approved it. I defer to @bgoodri on merging and CRAN release.

**Comment by hsbadr** [comment]
Sorry for the delay! I'll test this ASAP.

**Comment by andrjohns** [comment]
Apologies to push more changes, but the implementation was failing for cases where the plain type wasn't templated, so have fixed that.

This was an issue for the `lgpr` package, so you can also test against this fork: https://github.com/andrjohns/lgpr/tree/rstan-next

**Comment by andrjohns** [comment]
@hsbadr and @bgoodri I've updated this [discourse post](https://discourse.mc-stan.org/t/r-package-maintainers-using-stan-please-read/24131/14) to track the reverse-dependencies that are blocking the next release.

After this PR is merged and out to CRAN, there will be:
 
- 14 packages blocking rstan 2.21 & StanHeaders 2.26
- 8 packages additionally blocking rstan 2.26 & StanHeaders 2.26

All of which have either had PRs submitted or the maintainer contacted. Several packages will still fail on CRAN, but have already merged PRs with fixes into their dev repos.

I think once this PR gets out to CRAN, we could advertise a cut-off date for the above packages to merge the changes before we go ahead with submission. I think that would satisfy CRAN that we've done as much as possible

**Comment by hsbadr** [comment]
@andrjohns I agree with you that's sufficient and CRAN can push really hard to force the package maintainers to update their packages (or simply merge and release your PRs).

On a related note, I've tested two alternative solutions that work together but are probably not recommended. One redirects the Stan/Math headers into `rstan` to free it from `StanHeaders` dependency (so we can release updates directly by maintaining `rstan` only, mostly) -- This requires legal advice on licensing related issues. The second fixes the TBB dependency by building the required code into `libStanHeaders`. On one hand, this is not recommended by Intel, but on the other hand, it fixes many packages and will let us control TBB version directly from Math.

I personally vote for your approach (officially contact the maintainers to apply the required changes in the PRs & give them a couple weeks before submitting `StanHeaders v2.26.4` to CRAN and reporting the taken actions).

I'm tagging @wds15 b/c I think he's interested in this discussion.

**Comment by andrjohns** [comment]
@hsbadr I mentioned this PR to @bgoodri in the Stan meeting and he said that the `USE_STANC3` definition should have been handled by stanc, rather than needing the additional makeflags.

It turns out that `rstan_config()` was inserting the `#define USE_STANC3` _after_ `"#include <rstan/rstaninc.hpp>"`, which means that all of the rstan/stan headers were being included before the stanc3 flag was set. By updating `rstan_config` to append the `rstaninc` header after the definition ([relevant change here]("https://github.com/stan-dev/rstantools/pull/85/files#diff-c75c61b97a82db4f9119e131ea8d719aabc483e10f55e52ffa9248cc6dcc9ca5R216-R217"), packages are compiling fine with 2.26+2.26 without needed the additional makevars flag

**Comment by hsbadr** [comment]
@andrjohns Any code generated by the experimental version of `rstan` does define `USE_STANC3`, which can be removed when/if `stanc3` handles it. But, some packages may fail when they somehow customize the (order of) headers or have generated compatible C++ code by an earlier version of `rstan >= 2.26`.

https://github.com/stan-dev/rstan/blob/experimental/rstan/rstan/R/stanc.R#L218-L223

**Comment by bgoodri** [comment]
> The second fixes the TBB dependency by building the required code into `libStanHeaders`. On one hand, this is not recommended by Intel, but on the other hand, it fixes many packages and will let us control TBB version directly from Math.

I think if we can do something along the lines linking libStanHeaders.{so,dylib,dll} to the TBB shared library, then we could keep progressing on releases while pushing package maintainers to fix their linking and migrating toward rstantools 2.x style packaging.

**Comment by hsbadr** [comment]
> > The second fixes the TBB dependency by building the required code into `libStanHeaders`. On one hand, this is not recommended by Intel, but on the other hand, it fixes many packages and will let us control TBB version directly from Math.
> 
> I think if we can do something along the lines linking libStanHeaders.{so,dylib,dll} to the TBB shared library, then we could keep progressing on releases while pushing package maintainers to fix their linking and migrating toward rstantools 2.x style packaging.

The problem here is that the TBB shared library won't be in the library path for the packages, unless it's a system library or they're linking to `RcppParallel`. What I've tested, and it works, is automatically adding some/all of the TBB objects from Math into the static library `libStanHeaders.a` and that also allows shipping the TBB headers without version conflicts. So, basically, we treat TBB like `sundials`. I think we could exploit this workaround at least temporarily until we can revert it later. If you agree, I can merge the changes and update the `revdep` results.

**Comment by andrjohns** [comment]
> But, some packages may fail when they somehow customize the (order of) headers or have generated compatible C++ code by an earlier version of rstan >= 2.26.

Yeah I did run into that issue when doing some more testing, so reverted the makevars changes.

@bgoodri are you happy to merge these changes?

**Comment by bgoodri** [comment]
Let's try it.

**Review by bgoodri — APPROVED** [review]

**Comment by andrjohns** [comment]
Great, thanks! @jgabry would you be able to submit an updated version to CRAN when you have a minute? Once this hits CRAN we can put a clock on package maintainers to merge and release the PRs that I've opened for their packages before we submit the new StanHeaders

**Comment by bgoodri** [comment]
Let's see where we are at with Hamada's idea to treat TBB like Sundials in
StanHeaders before making any proclamations.

**Comment by andrjohns** [comment]
Isn't that approach only necessary while the reverse dependencies themselves aren't configured for linking to the TBB? We only have 14 packages that aren't setup for this yet, all of which I've opened PRs for ([more in this post](https://github.com/stan-dev/rstantools/pull/85#issuecomment-930644878))

**Comment by hsbadr** [comment]
> Let's try it.

Great! I'm on it.

> Isn't that approach only necessary while the reverse dependencies themselves aren't configured for linking to the TBB? We only have 14 packages that aren't setup for this yet, all of which I've opened PRs for ([more in this post](https://github.com/stan-dev/rstantools/pull/85#issuecomment-930644878))

True. But, as far as the changes in those PRs (specifically linking to `RcppParallel` and respecting its flags for library paths & `-ltbb`) aren't submitted to CRAN, our reverse dependency checks will fail.

**Comment by jgabry** [comment]
@andrjohns @hsbadr @bgoodri I'm now getting an error from `rstan_config` that we didn't used to get.  One way to reproduce is to try knitting the vignette "minimal-rstan-package.Rmd". When I do that I get: 

```
Error in if (!after) c(values, x) else if (after >= lengx) c(x, values) else c(x[1L:after],  : 
  argument is of length zero
```

coming from `rstan_config`. I also get this error from calling `rstan_config` outside the vignette after using `rstan_create_package`, adding a Stan file, then calling `rstan_config`.

**Pull request merged**